### ISSUE TITLE
[RFC] SDPA API Changes

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#110681]
# Updated SDPA API

**Authors:**
* @drisspg
## **Summary**
In order for users to more easily manage the complexity handling of various masking formats we would like to expose the ability to pass in `AttnMask` derived classes to SDPA and have it dispatch to the most optimal kernel.

## **Motivation**

Managing various mask types and their specializations in SDPA can become intricate for users. By enabling them to pass in custom mask types, we aim to streamline this process. Furthermore, this change would empower features like the `score_mod` API. Such features have shown considerable performance improvements when used in conjunction with `torch.compile` and SDPA, especially for specific mask types.

This would also enable PyTorch to upgrade its FlashAttention Kernel to the Newest Implementation. We are currently limited by BC concerns, see: https://github.com/pytorch/pytorch/issues/108108.
And user motivation for this change: https://github.com/pytorch/pytorch/issues/110144

## **Proposed Implementation**:
The foundational changes to the C++ core are outlined in this PR: https://github.com/pytorch/pytorch/pull/110399.
A prototype implementation can be found in this GitHub repository: https://github.com/drisspg/transformer_nuggets/pull/5/. This Prototype implementation would be incorporated into the existing top level nn.functional version of SDPA. 

Currently, the interface for invoking the `scaled_dot_product_attention` looks like:

`scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, *, float? scale=None) -> Tensor`

These changes would update the signature of sdpa to instead of accepting an `attn_mask: Tensor` accept a derived class from 
``` Python
class AttnMask(ABC):
    """Abstract base class for attention masks"""

    @abstractmethod
    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
        raise NotImplementedError("This is an abstract base class")

    @abstractmethod
    def needs_materialization(self) -> bool:
        raise NotImplementedError("This is an abstract base class")

    @staticmethod
    @abstractmethod
    def dispatch(
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
        attn_mask: "AttnMask",
        causal: bool,
        scale: Optional[float],
        dropout_p: float,
    ) -> torch.Tensor:
        raise NotImplementedError("This is an abstract base class")

```

To ensure that existing functionality remains uninterrupted, we would also introduce popular derived classes, such as:
``` Python
class TensorMask(AttnMask):
    """A mask that is a tensor"""

    def __init__(self, mask: torch.Tensor):
        self.mask = mask

    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
        return self.mask

    def needs_materialization(self) -> bool:
        return True
    
    @staticmethod
    def dispatch(
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
        attn_mask: "TensorMask",
        causal: bool,
        scale: Optional[float],
        dropout_p: float,
    ) -> torch.Tensor:
        raise NotImplementedError(
            "TensorMask requires materialization, so this should never be called!"
        )
    	    
```

#### CausalMask Variants
An example that would utilize all features of the AttnMask class can be found [Here](https://github.com/drisspg/transformer_nuggets/blob/sdpa_api/transformer_nuggets/sdpa/attn_mask.py#L93-L207)
#### Changes to torch.nn.functional.scaled_dot_product_attention()
A quick summary along with implementation can be found below:
- If the passed `attn_mask` is a tensor, the old method of computation is followed, along with a warning.
- If no `attn_mask` is provided, the function computes without any mask.
- If an `attn_mask` derived from `AttnMask` is provided and requires materialization (like `TensorMask`), the function materializes it and then computes.
- If an `attn_mask` derived from `AttnMask` doesn't need materialization, it uses the `dispatch` method of that mask class to handle the computation.

``` Python
def sdpa_prototype(
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attn_mask: Optional[AttnMask] = None,
    causal: bool = False,
    scale: Optional[float] = None,
    dropout_p: float = 0.0,
):
    assert attn_mask is None or isinstance(attn_mask, (AttnMask, torch.Tensor))

    if isinstance(attn_mask, torch.Tensor):
        warn("Passing a tensor as an attn_mask is deprecated. Please use TensorMask instead.")
        return scaled_dot_product_attention(query, key, value, attn_mask, dropout_p, causal, scale)

    if not attn_mask:
        return scaled_dot_product_attention(
            query, key, value, dropout_p=dropout_p, is_causal=causal, scale=scale
        )

    if attn_mask.needs_materialization():
        return scaled_dot_product_attention(
            query,
            key,
            value,
            attn_mask=attn_mask.materialize(query.device),
            dropout_p=dropout_p,
            is_causal=causal,
            scale=scale,
        )
    # After this point all AttnMask are required to have defined their own dispatching logic
    return attn_mask.dispatch(query, key, value, attn_mask, causal, scale, dropout_p)
```

## **Metrics**
- **Performance** Measure the performance improvement for popular additive masks like Alibi
- **Kernel Dispatch Overhead**: Quantify the overhead introduced due to the new dispatch mechanism. Ideally, this overhead should be negligible.
- **API Adoption Rate**: Track how many developers switch to using the new API over a specified time frame. A higher adoption rate signifies a positive reception.
- **User Feedback**: Collect qualitative feedback from developers. This can provide insights into the usability and potential areas of improvement.

## **Drawbacks**
Please consider:
- This is not immediately user breaking but we would enable a deprecation cycle for passing in regular tensors.
- I think this would meaningfully improve the UX but would also introduce some complexity
- This would add a non trivial amount of new code to an existing HIGHLY exercised code path, and with that could introduce new bugs. However part of this changes enables much more of the logic to be completely in Python which users find much more understandable and easier to debug then C++. 
## Alternatives

- **Introducing Specialized Functions**: 
  - One approach is to introduce specialized functions similar to `scaled_dot_product_attention`. An example could be `sdpa_causal_lower_right()`.
  - However, this could lead to challenges in maintainability and potentially complicate the user experience.

- **Expanding Existing Function with More Arguments**: 
  - We could enhance the primary `SDPA` function by adding more arguments to accommodate the desired features.
  - This could result in a bloated function signature reminiscent of `nn.mha`, which could overwhelm users with a plethora of keyword arguments.

- **Tensor Subclasses**
	-  This is somewhat analogous to tensor_subclass. We could define `__torch_function__` for the individual classes in question and overload the logic for calling into SDPA. Part of me feels like this is more machinery then needed. And I have concerns with the handling with torch.compile. That being said I do think this is the most attractive alternative to the proposed implementation. 
## **Prior Art**
This implementation is heavily borrowed/inspired from Xformers. They have a notion of an [AttnMask](https://github.com/facebookresearch/xformers/blob/3dafaf41aa3ae672057c51b96aba79c38a26ac16/xformers/components/attention/attention_mask.py#L14).I think that the communities reception of this technique is generally positive. And enjoy this abstraction. 
## **How we teach this**
* What names and terminology work best for these concepts and why? How is this idea best presented?
	* I think by updating the function documentation and providing examples of how users can interact with a kvcache using the causal variants in a tutorial would be helpful. Similar to what this user was asking for here: https://github.com/pytorch/pytorch/issues/110144
* Would the acceptance of this proposal mean the PyTorch documentation must be re-organized or altered? Yes
* How should this feature be taught to existing PyTorch users? I think the hardest thing to teach would be the lambda mask but I think that its design/implementation can be independent of this proposal and would likely be a feature on its own that this API change would enable.

## Unresolved Questions

- **Design Clarifications Through the RFC Process**: 
  - One area of potential improvement is the coexistence of both the `causal` flag and the `AttnMask`. This might inadvertently reveal too much about the kernel's implementation to users.
  - While there is an inclination towards retaining this due to its potential utility with the `score mod` API, it warrants further discussion and consensus.

- **Design Resolutions Before Stabilization**: 
  - Feedback is sought on the envisioning of user-defined masks. Specifically, how they might be structured and if the proposed API is versatile enough to encompass them.

**Comment by drisspg** [comment]
Related: https://github.com/pytorch/pytorch/issues/110702

**Comment by scaomath** [comment]
Adding this would be a nice upgrade for SDPA research.

**Comment by drisspg** [comment]
## __torch__ dispatch__ Tensor Subclass and torch.scaled_dot_product_attention()
When attempting to pass a __dispatch__ tensor subclass as an attn_mask to the torch.scaled_dot_product_attention() function, we encounter a compilation issue. The primary reason behind this is the composite nature of torch.scaled_dot_product_attention(). Due to its composite design, its internal operations are abstracted, making it challenging to directly pattern-match for __torch__dispatch__.

Potential matchable patterns within this function could include _sdpa_math, _sdpa_mem_eff, and _sdpa_flash. Although users can override these, the full dispatch process in C++ gets triggered, which might not be implicitly composite. A plausible workaround might be to:

Check if the tensor is a Python subclass using `is_python_dispatch()`.
If true, short-circuit this check to a predefined dummy operation.
Allow users to override this dummy operation for dispatch purposes.
## __torch_function__ Compilation Issue
Currently, there is a limitation where torch_function cannot be compiled. This limitation restricts its application in certain scenarios.

cc @mlazos curious what your take would be on this

**Comment by danthe3rd** [comment]
Hi,
That's a very interesting proposal! A few comments from our experience in `xformers` with that:
1. The `AttnMask` class you quoted in xFormers is actually deprecated and quite complicated. The ones we support moving forward are there: https://github.com/facebookresearch/xformers/blob/main/xformers/ops/fmha/attn_bias.py
2. I believe it makes more sense to refer to these objects as Attention Biases rather than Masks, because it's simply more generic (Mask = Bias with -inf), and Biases support things like alibi
3. Regarding the `causal` flag, it could also be a sort of bias
4. Regarding the `materialize` function, in xformers it takes shape, dtype and device arguments. This allows to create a causal mask with simply `LowerTriangularMask()` and not care about shapes at this time
5. Something worth thinking about: maybe biases should have a `to` method to put the bias on the right device. Ideally the user should be responsible for doing that, rather than the `dispatch` or implementation
6. The `dispatch` function adds a lot of complexity to this class, which not only represents the mask, but also how it should be used. In xFormers, each operator/backend [has a list of supported masks](https://github.com/facebookresearch/xformers/blob/9592c04d9a7b9b939e51547cf9ffe250489d8f65/xformers/ops/fmha/flash.py#L353) it can accept, and the SDPA API dispatch could just be to rank operators by speed, and take the first one that supports this mask. This still raises the question of how to support extending with custom biases that require custom backends tho...

cc @fmassa @bottler

**Comment by drisspg** [comment]
@danthe3rd  Thanks for the comments I really appreciate it!

2. You are totally right and going to update this RFC

3.  Also agree and this is more an artifact of having to be non BC breaking, and indeed I write causal bias as an attn_bias [here](https://github.com/drisspg/transformer_nuggets/blob/ce0afa0f4e9eb57578453a31380d0ab8f3a85733/transformer_nuggets/sdpa/attn_bias.py#L100C28-L100C28 ). I think we may want to deprecate the is_causal flag going forward. However in the context of the [score mod api](https://docs.google.com/document/d/1yDE16Tv57Sk7tMC3QFZ7bJ4BgkiFMIUt_WVtL8FwVKY/edit#heading=h.mgr8aoftnhr6) having a strong causal signal is beneficial for populating the template. That could also be rolled up into the attn-bias though.

4. Hmm this is interesting, is there strong reason for delaying this? Do you by chance have an example of where the shapes wouldn't be known at mask construction but only known at materialization. Im guessing this might be beneficial for having one mask that is passsed into every attn block and is not recreated every iteration for varying seq_lens.

5. Also a good point, I added a device in the `materialize` function thinking that this would cover the use case. But that makes sense it might be nice having the ability to do this independent of materialization

6. This indeed is the crux of it lol

**Comment by fmassa** [comment]
Hi,

My 2 cents here (and which goes a bit differently than what xFormers implement) is that I'd consider the `AttentionBias` as imply a Tensor subclass. This way, part of the dispatching logic would be handled by PyTorch already (when the user creates a new AttentionBias, for example for block-diagonal etc), and the only additional dispatching needed would be to pick the fastest kernel for a given AttentionBias.

And I'd pass the shapes / dtypes during construction time, and instead of having a `materialize` function in the Tensor subclass reuse something like `Tensor.to_dense`.

That being said, we should check if this is the first occurence of such use of Tensor subclasses in mainstream PyTorch, as if the functionality isn't really ready (as was the case a [couple of years ago when I first tried it](https://github.com/facebookresearch/xformers/blob/main/xformers/sparse/blocksparse_tensor.py#L100) then it would be quite frustrating to users

**Comment by vince62s** [comment]
Would that make sense to add the equivalent of `flash_attn_with_kvcache` in SDPA ?
The main benefit when providing a KV cache that is modified in place is the speed.

**Comment by drisspg** [comment]
Yup, I plan on working on this

**Comment by samvanstroud** [comment]
Flash attention recently added support for local window attention. Would it be possible to expose this functionality via the pytorch SDPA interface?

**Comment by vince62s** [comment]
> Would that make sense to add the equivalent of `flash_attn_with_kvcache` in SDPA ? The main benefit when providing a KV cache that is modified in place is the speed.

any ETA on this? I mean in terms of pytorch version. thanks.

**Comment by vince62s** [comment]
will this be handled along with this: https://github.com/pytorch/pytorch/issues/121465 ?

**Comment by vadimkantorov** [comment]
Also, the docs reference doesn't seem to explain the conditions needed to run flash attention (e.g. the question is will it run if we pass an arbitrary relpos attn mask):
- https://pytorch.org/docs/stable/backends.html#torch.backends.cuda.can_use_flash_attention
- https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch-nn-functional-scaled-dot-product-attention

**Comment by npuichigo** [comment]
Does flash attention kernel support custom attn_mask?

**Comment by vadimkantorov** [comment]
On a more general question of custom `attn_bias`: I think flash_attn does not currently support this, but there is a very draft PR on this theme:
- https://github.com/Dao-AILab/flash-attention/pull/617
Probably this PR/something similar would need to be completed to have custom attn_bias support

About attn_mask - not sure, maybe there is more specialized support for this already merged...

**Comment by vadimkantorov** [comment]
@drisspg Would FlexAttention be a faster backend than SDPA for a passed attn_bias (than existing mem_efficient)?

**Comment by drisspg** [comment]
@vadimkantorov 
- I have a gist exploring the difference https://gist.github.com/drisspg/e7ce84a3ecc7ea1a47b99b267a11e2a9
I get roughly a 1.5x speed up for FlexAttention over SDPA on my machine. The only path we have for full additive bias on SDPA today is memory-efficient attention which is much slower than FAv2. 

That being said @eqy has been working on enabling CuDNN backend for additive bias on H100 machines

- It depends on how you do VarLen sequence. We do have support for dynamic shapes: https://github.com/pytorch/pytorch/blob/acad2050c14cf7585e210d49d72789e3d0f295e5/test/inductor/test_flex_attention.py#L356

But if you do sequence stacking similar to what is mentioned about the document mask in the blog your shapes would be static.

And yes it is possible to dump the triton code

**Comment by vadimkantorov** [comment]
> I get roughly a 1.5x speed up for FlexAttention over SDPA on my machine. 

Oh wow. Hope it gets added as a backend to SDPA interface as well :)

**Comment by vince62s** [comment]
@drisspg 
> memory-efficient attention which is much slower than FAv2

in what context ? for short (up to 256/512 tokens) sequences memory-efficient is as fast as FAv2 and even faster in my use case.
